### PR TITLE
tokenize empty strings

### LINKED ISSUE #76: tokenize empty string
Currently when we try to tokenize an empty string, it throws an error an out of bounds error.
```python
doc = nlp("")
# throws an error
```
@Nilanshrajput can you look into this issue as you worked on the tokenizer recently? Maybe create an empty doc in this case.

### PR DESCRIPTION
**Pull request by Nilanshrajput**
# Pull Request Template

## Description

return empty doc([]) for empty strings ("")

Fixes #76 

## Type of change

Please mark options that are relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by Nilanshrajput** [comment]
@AlanAboudib @bicycleman15  why is this error coming?
 'PointerTensor' object has no attribute 'get_encrypted_vector'

**Comment by bicycleman15** [comment]
One thing, I don't think we need the `text` argument anymore. you can remove that.

**Comment by bicycleman15** [comment]
> @AlanAboudib @bicycleman15 why is this error coming?
> 'PointerTensor' object has no attribute 'get_encrypted_vector'

When does it come ?

**Comment by Nilanshrajput** [comment]
> > @AlanAboudib @bicycleman15 why is this error coming?
> > 'PointerTensor' object has no attribute 'get_encrypted_vector'
> 
> When does it come ?

on call of doc_pinter.get_encrypted_vector() in tutorial part 1. have at look at ci build

**Comment by Nilanshrajput** [comment]
> One thing, I don't think we need the `text` argument anymore. you can remove that.

text from tokenizer call? I think it's needed

**Comment by bicycleman15** [comment]
> > One thing, I don't think we need the `text` argument anymore. you can remove that.
> 
> text from tokenizer call? I think it's needed

I removed it in the last PR I made as we were storing hashes of strings so no need to store text now in doc. Look at #74

**Comment by bicycleman15** [comment]
Idk why `DocPointer` is getting converted to `PointerTensor`. In the tutorial example, `doc_ptr` should be a `DocPointer` instead of a `PointerTensor`. Maybe manually check in a notebook with your own example.

**Comment by Nilanshrajput** [comment]
> > > One thing, I don't think we need the `text` argument anymore. you can remove that.
> > 
> > 
> > text from tokenizer call? I think it's needed
> 
> I removed it in the last PR I made as we were storing hashes of strings so no need to store text now in doc. Look at #74

Don't worry the text is not saved in doc. the argument is in tokenizer it's not stored anywhere!

**Comment by AlanAboudib** [comment]
When you see `PointerTensor` this indicates that creation of the `DocPointer` is not working. I should check this out

File: syfertext/tokenizer.py [hunk 1]
```diff
@@ -138,11 +138,15 @@ def __call__(self, text: Union[String, str]):
         # I do not assign the Doc here any owner, this will
         # be done by the SupPipeline object that operates
         # this tokenizer.
-        doc = Doc(self.vocab)
+        doc = Doc(self.vocab, text)
```
Comment (AlanAboudib) [inline syfertext/tokenizer.py:141]: Why `text` is in the constructor?

File: syfertext/tokenizer.py [hunk 2]
```diff
@@ -138,11 +138,15 @@ def __call__(self, text: Union[String, str]):
...
 
         # The number of characters in the text
         text_size = len(text)
```
Comment (AlanAboudib) [inline syfertext/tokenizer.py:144]: no need for this line, you can use `len(text)` directly in the if statement. it is pretty clear.

File: syfertext/tokenizer.py (hunk 1 above)
Comment (Nilanshrajput) [reply syfertext/tokenizer.py:141]: I missed removing it. updated now

File: syfertext/tokenizer.py (hunk 2 above)
Comment (Nilanshrajput) [reply syfertext/tokenizer.py]: actually it was in use at 2 other places previously also!
anyway i updated them

**Comment by Nilanshrajput** [comment]
@AlanAboudib  the DocPointer error is not coming now somehow.

**Comment by AlanAboudib** [comment]
@Nilanshrajput :\

File: syfertext/tokenizer.py [hunk 3]
```diff
@@ -140,8 +140,9 @@ def __call__(self, text: Union[String, str]):
         # this tokenizer.
         doc = Doc(self.vocab)
 
-        # The number of characters in the text
-        text_size = len(text)
```
Comment (AlanAboudib) [inline syfertext/tokenizer.py:144]: I haven't noticed that `text_size` was used multiple times. Can you put it back pleae :'D

Comment (Nilanshrajput) [reply syfertext/tokenizer.py:144]: I kinda already know this was going to happen (:
updated

**Comment by Nilanshrajput** [comment]
> @Nilanshrajput :\

I guess the error was due to the text in doc constructor.

**Comment by sachin-101** [comment]
@Nilanshrajput @AlanAboudib  the doc owner is initialized on in the tokenizer and then after being processed by the Tokenizer it is assigned it's `owner` and `client_id` in the subpipeline (which the tokenizer is a part of).
Is it fine to return the doc object directly from the tokenizer, without the right attributes? At times, we may even need to return DocPointer).
I guess it is not very consistent and might break the user's code.

**Comment by AlanAboudib** [comment]
Can you explain what you mean by this @sachin-101 

> Is it fine to return the doc object directly from the tokenizer, without the right attributes? At times, we may even need to return DocPointer).

**Comment by sachin-101** [comment]
Oh ! Actually it is fine to return `Doc` from `tokenizer`, since it is the `subpipeline` which is calling the tokenizer, and after the call it is assigned the right `owner` and `client` in the subpipeline.

But I have another doubt now 😅, is it fine to pass the empty doc to all the pipeline components, can they handle this special case?
And can we avoid it (may be to save some time), by checking the doc, before feeding into other pipe components?

**Comment by Nilanshrajput** [comment]
@sachin-101  its fine i think, if any component try access empty doc, key error should be raised.

**Comment by AlanAboudib** [comment]
@sachin-101 This shouldn't be a problem, each pipe element should know how to handle an empty doc. I think this is not a very realistic usecase, if it happens, it should be extremely rare, so no real waste of time

**Pull request merged**